--- Day 4: Camp Cleanup ---

Space needs to be cleared before the last supplies can be unloaded from the ships, and so several Elves have been assigned the job of cleaning up sections of the camp. Every section has a unique ID number, and each Elf is assigned a range of section IDs.

However, as some of the Elves compare their section assignments with each other, they've noticed that many of the assignments overlap. To try to quickly find overlaps and reduce duplicated effort, the Elves pair up and make a big list of the section assignments for each pair (your puzzle input).

For example, consider the following list of section assignment pairs:

    2-4,6-8
    2-3,4-5
    5-7,7-9
    2-8,3-7
    6-6,4-6
    2-6,4-8

For the first few pairs, this list means:

Within the first pair of Elves, the first Elf was assigned sections 2-4 (sections 2, 3, and 4), while the second Elf was assigned sections 6-8 (sections 6, 7, 8).
The Elves in the second pair were each assigned two sections.
The Elves in the third pair were each assigned three sections: one got sections 5, 6, and 7, while the other also got 7, plus 8 and 9.
This example list uses single-digit section IDs to make it easier to draw; your actual list might contain larger numbers. Visually, these pairs of section assignments look like this:

    .234.....  2-4
    .....678.  6-8

    .23......  2-3
    ...45....  4-5

    ....567..  5-7
    ......789  7-9

    .2345678.  2-8
    ..34567..  3-7

    .....6...  6-6
    ...456...  4-6

    .23456...  2-6
    ...45678.  4-8

Some of the pairs have noticed that one of their assignments fully contains the other. For example, 2-8 fully contains 3-7, and 6-6 is fully contained by 4-6. In pairs where one assignment fully contains the other, one Elf in the pair would be exclusively cleaning sections their partner will already be cleaning, so these seem like the most in need of reconsideration. In this example, there are 2 such pairs.

In how many assignment pairs does one range fully contain the other?

In [27]:
import pandas as pd

In [54]:
txt = "27-42,11-60"
txt2 = "3-5,2-5"

def overlapping_range(left, right):
    left = list(map(int, left.split("-")))
    right = list(map(int, right.split("-")))

    left = list(range(left[0], left[1]+1))
    right = list(range(right[0], right[1]+1))

    intersection_length = len(set(left).intersection(set(right)))
    return len(left) == intersection_length or len(right) == intersection_length

print(overlapping_range(*txt.split(",")))
print(overlapping_range(*txt2.split(",")))

True
True


In [55]:
pd.read_csv("cleanup pairs example.txt", header=None, names=["left", "right"]).apply(lambda x: overlapping_range(x["left"], x["right"]), axis=1).sum()

2

In [56]:
pd.read_csv("cleanup pairs.txt", header=None, names=["left", "right"]).apply(lambda x: overlapping_range(x["left"], x["right"]), axis=1).sum()

424

First try: 442. Failed because my string approach made mistakes when a range had only 1 digit (e.g. 1-1). This way, 1-1 was being found in 11-12, which is wrong.
Will use a numeric approach instead.

Second try, 396 was too low. Sets are unorded and I had forgotten, so the comparison with the ranges was incorrect. Instead I ended up comparing length of the set with length of the ranges.

424 was the right answer!

--- Part Two ---

It seems like there is still quite a bit of duplicate work planned. Instead, the Elves would like to know the number of pairs that overlap at all.

In the above example, the first two pairs (2-4,6-8 and 2-3,4-5) don't overlap, while the remaining four pairs (5-7,7-9, 2-8,3-7, 6-6,4-6, and 2-6,4-8) do overlap:

    5-7,7-9 overlaps in a single section, 7.
    2-8,3-7 overlaps all of the sections 3 through 7.
    6-6,4-6 overlaps in a single section, 6.
    2-6,4-8 overlaps in sections 4, 5, and 6.
    So, in this example, the number of overlapping assignment pairs is 4.

In how many assignment pairs do the ranges overlap?

In [62]:
txt = "27-42,43-60"
txt2 = "3-3,3-5"

def overlapping_range(left, right):
    left = list(map(int, left.split("-")))
    right = list(map(int, right.split("-")))

    left = list(range(left[0], left[1]+1))
    right = list(range(right[0], right[1]+1))

    return bool(set(left).intersection(set(right)))

print(overlapping_range(*txt.split(",")))
print(overlapping_range(*txt2.split(",")))

False
True


In [64]:
pd.read_csv("cleanup pairs example.txt", header=None, names=["left", "right"]).apply(lambda x: overlapping_range(x["left"], x["right"]), axis=1).sum()

4

In [63]:
pd.read_csv("cleanup pairs.txt", header=None, names=["left", "right"]).apply(lambda x: overlapping_range(x["left"], x["right"]), axis=1).sum()

804